src/core/p5.Renderer.js

```diff
@@ -35,69 +40,107 @@ p5.Renderer = class Renderer extends p5.Element {
+    this.states = {
+      doStroke: true,
+      strokeSet: false,
+      doFill: true,
+      fillSet: false,
+      tint: null,
+      imageMode: constants.CORNER,
+      rectMode: constants.CORNER,
+      ellipseMode: constants.CENTER,
+      textFont: 'sans-serif',
+      textLeading: 15,
+      leadingSet: false,
+      textSize: 12,
+      textAlign: constants.LEFT,
+      textBaseline: constants.BASELINE,
+      textStyle: constants.NORMAL,
+      textWrap: constants.WORD
+    };
+    this.pushPopStack = [];
+
+
+
     this._clipping = false;
     this._clipInvert = false;
 
-    this._textSize = 12;
-    this._textLeading = 15;
-    this._textFont = 'sans-serif';
-    this._textStyle = constants.NORMAL;
-    this._textAscent = null;
-    this._textDescent = null;
-    this._textAlign = constants.LEFT;
-    this._textBaseline = constants.BASELINE;
-    this._textWrap = constants.WORD;
-
-    this._rectMode = constants.CORNER;
-    this._ellipseMode = constants.CENTER;
+    // this._textSize = 12;
+    // this._textLeading = 15;
+    // this._textFont = 'sans-serif';
+    // this._textStyle = constants.NORMAL;
+    // this._textAscent = null;
+    // this._textDescent = null;
+    // this._textAlign = constants.LEFT;
+    // this._textBaseline = constants.BASELINE;
+    // this._textWrap = constants.WORD;
+
+    // this._rectMode = constants.CORNER;
+    // this._ellipseMode = constants.CENTER;
     this._curveTightness = 0;
-    this._imageMode = constants.CORNER;
+    // this._imageMode = constants.CORNER;
 
-    this._tint = null;
-    this._doStroke = true;
-    this._doFill = true;
-    this._strokeSet = false;
-    this._fillSet = false;
-    this._leadingSet = false;
+    // this._tint = null;
+    // this._doStroke = true;
+    // this._doFill = true;
+    // this._strokeSet = false;
+    // this._fillSet = false;
+    // this._leadingSet = false;
 
     this._pushPopDepth = 0;
   }
 
+  id(id) {
+    if (typeof id === 'undefined') {
+      return this.elt.id;
+    }
+
+    this.elt.id = id;
+    this.width = this.elt.offsetWidth;
+    this.height = this.elt.offsetHeight;
+    return this;
+  }
+
   // the renderer should return a 'style' object that it wishes to
   // store on the push stack.
   push () {
     this._pushPopDepth++;
-    return {
-      properties: {
-        _doStroke: this._doStroke,
-        _strokeSet: this._strokeSet,
-        _doFill: this._doFill,
-        _fillSet: this._fillSet,
-        _tint: this._tint,
-        _imageMode: this._imageMode,
-        _rectMode: this._rectMode,
-        _ellipseMode: this._ellipseMode,
-        _textFont: this._textFont,
-        _textLeading: this._textLeading,
-        _leadingSet: this._leadingSet,
-        _textSize: this._textSize,
-        _textAlign: this._textAlign,
-        _textBaseline: this._textBaseline,
-        _textStyle: this._textStyle,
-        _textWrap: this._textWrap
-      }
-    };
+    const currentStates = Object.assign({}, this.states);
```

Comment:
This should probably be using `structuredClone()` because states may have objects as properties that won't be fully preserved if copied in this way. It is currently not `structuredClone()` because somewhere in the WebGL renderer states is a reference to `window` which cannot be cloned.

Reply:
I believe only [some base types](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Structured_clone_algorithm#javascript_types) work in structured clone, and I'm not certain all the things we put into this state conform to that (e.g. in WebGL, currently we have a camera object.) We have a few options here, do you have any preference between:
- making an interface like `clone()` that state properties can optionally implement so that we can have custom object types in the state
- forcing all state to be a base type and not storing custom classes

I think the former might be a smaller refactor, but that extra flexibility might also mean a higher surface area for bugs. If we go for the latter, for WebGL, I think that would mean storing camera properties rather than the camera itself, and adding a custom handler on `pop()` to reapply those properties to the active camera.

Reply:
I think for performance we might want the option of not deep copying everything if it isn't necessary, so I maybe slightly lean towards the former?

Reply:
Some of the key things not cloneable for our use will be (including nested in objects) symbols, functions, and DOM nodes.

> making an interface like clone() that state properties can optionally implement so that we can have custom object types in the state

Does this mean we will loop through each direct properties of the `state` object and if there is a method `clone()` call it and store that?

> forcing all state to be a base type and not storing custom classes

I'm slightly more in favor of this and I think it is similar to the above option (if I understand it correctly) but with the renderer itself doing the `clone()` step before setting it in the state object.

For something like `p5.Camera` it was never suited for deep cloning anyway as it would be duplicating the underlying p5 instance and what not. Saving the camera properties then reapplying them to the active camera feels more semantic but I'm not sure if there will be any performance impact (through repeated instantiation of heavy object for example)?

Another option is that the base `p5.Renderer.states` object is essentially a convenience in that the push and pop stack will be implemented already in the base class so the derived class don't need to implement it, but if the derived class needs more complex behavior including using non-structuredCloneable states, it can perhaps be handled independently in the derived class's `push()` `pop()` implementation, if that made sense.

Reply:
> Does this mean we will loop through each direct properties of the `state` object and if there is a method `clone()`, call it and store that?

Yep!


> Another option is that the base `p5.Renderer.states` object is essentially a convenience in that the push and pop stack will be implemented already in the base class so the derived class don't need to implement it, but if the derived class needs more complex behavior including using non-structuredCloneable states, it can perhaps be handled independently in the derived class's `push()` `pop()` implementation, if that made sense.


My comment about perf was mostly about avoiding automatic deep copying of everything, so I'm good with this too if the more expensive clones are done explicitly when needed. So maybe this method would mean that by default, everything is a simple type that should be replaced instead of mutated in place so we can keep doing `Object.assign(...)` instead of a `structuredClone(...)`, and then a renderer can do whatever copy it needs manually in its `push` override? And it should be just a few things needing that custom behaviour.

Reply:
If we can further limit state properties to be basically primitives only (copy types only), `Object.assign()` should be enough, it is just when the state properties are objects or arrays then `structuredClone()` would be needed. I'm fine going with the primitives only approach for an overall lighter state management by default if there's no obvious need for reference types to be used as state properties?

If that sounds good, I can try to have a look at WebGL renderer states or since there are other WebGL stuff going on as well, maybe you'd have a better idea of when to review them? I'm going to move back onto the hierachy refactor and instance states next so it's not immediately blocking at this point.

Reply:
hmm right, I guess we have a lot of array states currently in WebGL as that's the format colors are stored in, and those are all treated as reference types right now. To avoid deep copies on all those, we'd have to wrap them in some other class to avoid the structured clone, so it feels like maybe that'd add more overhead than it's worth. Are there many 2D properties that need a clone? Would relying on object reference for everything by default be a possibility?

I tried doing some quick find-and-replacing this morning to start converting state into WebGL here: https://github.com/limzykenneth/p5.js/compare/2.0-modules...davepagurek:p5.js:2.0-webgl?expand=1 This is currently using the `.copy()` method on things if it exists, but we can change how that's done based on what we decide here, and currently this changeover has broken more tests that will need looking into. But anyway since this is just changing how some properties are accessed, I think it's ok to make these changes whenever, and it won't be too big of a merge for the other contributors.

Reply:
For 2D the only one using reference type is `tint` using an array of colors as far as I remember, but it is actually currently on the base p5.Renderer class as WebGL can tint as well. If there is either a guarantee or just enforcement through documentation that if reference types are set as state properties, it must not be separately modified, eg. additional tints creates new array and never reuses it, it would still be fine to just `Object.assign()`.

You should be able to push to this PR directly if you want to do the conversion and test fixing as necessary. I'll merge from dev-2.0 regularly as I work on this too.

Reply:
Btw for now I've got it still slicing arrays and calling `.clone()` on classes that support that because there were a lot more instances than I remembered. Still potentially feasible to change, but maybe not the top priority either. The updates to the math library will also affect how exactly cloning matrix properties works, so maybe that would be a good time to revisit and maybe refactor.